k9mail/src/main/java/com/fsck/k9/controller/QresyncSyncInteractor.java

```diff
@@ -0,0 +1,128 @@
+        this.listener = listener;
+        this.controller = controller;
+        this.imapSyncInteractor = imapSyncInteractor;
+    }
+
+    int performSync(QresyncParamResponse qresyncParamResponse, List<String> expungedUids, MessageDownloader messageDownloader)
+            throws MessagingException, IOException {
+        final List<ImapMessage> remoteMessages = new ArrayList<>();
+
+        if (account.syncRemoteDeletions()) {
+            List<String> deletedUids = new ArrayList<>(expungedUids);
+            deletedUids.addAll(qresyncParamResponse.getExpungedUids());
+            imapSyncInteractor.syncRemoteDeletions(deletedUids);
+        }
+
+        findNewRemoteMessagesToDownload(remoteMessages, qresyncParamResponse, messageDownloader);
+        if (imapFolder.getMessageCount() >= localFolder.getVisibleLimit() && imapFolder.getMessageCount() >=
+                (remoteMessages.size() + localFolder.getMessageCount())) {
+            findOldRemoteMessagesToDownload(remoteMessages);
+        }
+
+        return messageDownloader.downloadMessages(account, imapFolder, localFolder, remoteMessages, false, true, false);
+    }
+
+    private void findNewRemoteMessagesToDownload(List<ImapMessage> remoteMessages, QresyncParamResponse qresyncParamResponse,
+            MessageDownloader messageDownloader) throws MessagingException {
+
+        String folderName = imapFolder.getName();
+        final AtomicInteger headerProgress = new AtomicInteger(0);
+
+        for (MessagingListener l : controller.getListeners(listener)) {
+            l.synchronizeMailboxHeadersStarted(account, folderName);
+        }
+
+        List<ImapMessage> modifiedMessages = qresyncParamResponse.getModifiedMessages();
+        Timber.v("SYNC: Received %d FETCH responses in QRESYNC response", modifiedMessages.size());
+
+        String uid = localFolder.getSmallestMessageUid();
+        long smallestLocalUid = 1;
+        if (uid != null) {
+            smallestLocalUid = Long.parseLong(uid);
+        }
+
+        for (ImapMessage imapMessage : modifiedMessages) {
+            long remoteMessageUid = Long.parseLong(imapMessage.getUid());
+            Message localMessage = localFolder.getMessage(imapMessage.getUid());
+
+            if (remoteMessageUid < smallestLocalUid) {
+                continue;
+            }
+
+            if (localMessage == null || (!localMessage.isSet(Flag.X_DOWNLOADED_FULL) &&
+                    !localMessage.isSet(Flag.X_DOWNLOADED_PARTIAL))) {
+                remoteMessages.add(imapMessage);
+            } else {
+                messageDownloader.processDownloadedFlags(account, localFolder, imapMessage, headerProgress,
+                        modifiedMessages.size());
+            }
+        }
+
+        Timber.v("SYNC: Received %d new message UIDs in QRESYNC response", remoteMessages.size());
+    }
+
+    private void findOldRemoteMessagesToDownload(List<ImapMessage> remoteMessages) throws MessagingException {
```

Comment:
For QRESYNC, old messages are only downloaded if needed, and are specifically chosen. For the CONDSTORE/plain IMAP case, we take every message in the N most recent messages of the remote mailbox, and filter it down from there.